### PR TITLE
Implement save management, closes #40

### LINKED ISSUE #40: Feature Request - Save Management for DS/3DS Cartridges
Hello again!

I have a feature request and I'm unsure of where else to place this. I've noticed that GodMode9 has full support for dumping the ROM file in either full or trimmed format for DS cartridges instead of only 3DS cartridges, but one thing I'd like to have implemented is the ability to dump and replace the save data on the cartridge. A good place to look at for reference is [TuxSH's TWLSaveTool](https://github.com/TuxSH/TWLSaveTool) homebrew, where the feature is fully implemented as its own app in userland and also installable as an app for the system.

Thank you for all of your hard work once again!
If this cannot be done, all I ask is for an explanation as to why. However if it is just something you don't wish to focus on right now, that's also fully understandable and I wish you good luck with all else!

### PR DESCRIPTION
**Pull request by dratini0**
I have made some progress with the save management feature for #40 . Here is what I have so far:
* [x] Read/Write saves on NTR carts, with code borrowed from TWLSaveTool (and from Fastboot3DS)
* [x] Read/write encrypted saves on CTR carts by extending the code from TWLSaveTool

I would still like to do the following things before merging:
* [x] Test the code I have for Art Academy (FLASH_8MB)
* [ ] Test with a cartridge with infrared in it
* [x] Format the borrowed code to match the rest of GodMode9
* [x] Get rid of the old EEPROM.c

It would be nice to have the following:
* [ ] Decrypting the CTR saves
* [ ] Figuring out card2 saves (as far as I know, it is read just fine, as part of the untrimmed ROM, but we can't write it. It would also make sense to get it out into a separate file.)
* [ ] Figuring out what is going on with WarioWare DIY (that one has been bugging me!) [See here](https://github.com/TASVideos/desmume/blob/23f4dcc0094e9dcb77494593831b6aef9aaf3b5b/desmume/src/addons/slot1_retail_nand.cpp)

Any input on the code so far would be appreciated.

### CODE REVIEW

**Comment by Wolfvak** [comment]
Okay, wow, that's a lot of changes.
Gotta say, thanks for starting this on your own! Now, onto the problems...

- There's a lot of repeated/dead code. why is there a type for every single type of flash there's been? Could it not just be condensed to a few types (9bit EEPROM, 16bit EEPROM, 24bit FLASH, ...), and have a simple size or pagecount variable? GM9 is already pretty big as it is, so the least amount of unneeded code/data that's added, the better.

- The SPICARD code is essentially a duplicate of the same code already present for the ARM11. I was planning to have both processors share the same source code by putting it into the `common` folder, just like the PXI code (that's what https://github.com/d0k3/GodMode9/commit/f74c9bbf4b65d9f5d9a27fb1b06e90f9e233e446 was for).

The first point mostly applies to NDS carts - dunno if 3DS carts added so many different flash types where it actually becomes saner to include function ops for each type before trying to code a routine that can handle anything.

Other than that, this looks pretty darn good.

**Comment by d0k3** [comment]
Okay, nice!

But, try to make a clean pull request, don't do a pull request with merge conflicts (even resolved ones). And, as @Wolfvak said, the code may still need some cleanup, don't duplicate code.

**Comment by dratini0** [comment]
@Wolfvak :
1: I thought it would make sense to have an object for every type of flash there was in an NTR cart, as they are a bloody mess. Using block write on a 256K Sanyo is incredibly slow, while not so much on the ST chips. Then there is the mysterious 8M chip inside Art Academy, which likely doesn't support page write at all. In contrast, CTR flash chips (at least in card1 titles) are pretty regular, I have only ever seen three chips in one series. (Out of which one happens to actually be officially documented). What would make sense to me is getting rid of the CTR save types not sighted in the wild, and making infrared orthogonal to the actual save type (as in `struct CardType { const CardTypeData *chip; bool infrared; }`)

2: Good point.

@d0k3: Would rebasing it upon master work? Squashing it into one commit doesn't seem right.

**Comment by Wolfvak** [comment]
Merged SPI driver looks okay to me. I'd still like to look a bit further into savegame chip management, but that's one of two problems solved.

**Comment by d0k3** [comment]
@dratini0 - yeah, maybe rebasing on master works. We just don't want a borked commit history. You can also redo my "historical" commits yourself if you want to.

**Comment by aspargas2** [comment]
So, I tested a few carts building from commit 5b757f0, here's what I found:

* Mariokart DS works perfectly: dumped saves match with those form Checkpoint, and changing a save then restoring it reflects the change when dumping again with Checkpoint.
* Pokemon Black, Pokemon Omega Ruby, and Pokemon X (the foremost of which has builtin IR) all exhibit the same behavior: no `.sav` file appears in the virtual gamecart drive. The 4 byte `jedecid_and_sreg.bin` file does appear, but attempting to open the file action menu on it or copy it to anywhere yields a data abort. The exception dump for each of these carts is attached to this comment.
* Pokemon Black 2, another IR cart, appears to have inconsistent behavior: most of the time it exhibited the same behavior described in the previous point (its exception dump is also attached), but one time the `.sav` file actually did appear in the gamecart mount, and `jedecid_and_sreg.bin` was readable. However, the `.sav` file was only 512 bytes, and did not show any semblance of the actual save of the game when opened in the hex editor. I was unable to replicate this latter behavior, and it only occurred once, seemingly randomly, when mounting the virtual cart drive.
* Mariokart 7, Tetris Axis, Rayman 3D, Super Mario 3D Land, and Paper Mario Sticker Star all seem to work in that the `.sav` file properly appears and is readable, but I did not do any testing beyond that with them. If I ever do, I will comment again with an update.

All of this testing was done on a New 2DS XL, if that's at all relevant.

[exception_dump_pkmn_black2.txt](https://github.com/d0k3/GodMode9/files/3438482/exception_dump_pkmn_black2.txt)
[exception_dump_pkmn_oras.txt](https://github.com/d0k3/GodMode9/files/3438483/exception_dump_pkmn_oras.txt)
[exception_dump_pkmn_x.txt](https://github.com/d0k3/GodMode9/files/3438484/exception_dump_pkmn_x.txt)
[exception_dump_pkmn_black.txt](https://github.com/d0k3/GodMode9/files/3438485/exception_dump_pkmn_black.txt)

**Comment by dratini0** [comment]
@aspargas2 :
Amazing, thank you for the data. I think I have an idea why Omega Ruby and X crash: IIRC they are card2 titles (not yet supported), and I messed up the NO_CHIP case. Ooops.

For black and black2, I will have to investigate.

@Wolfvak : Yup, working on that.

@d0k3 : Thank you, I will do that.

**Comment by dratini0** [comment]
@aspargas2 : 
I think I have fixed the crashes now, it should be possible to read the JEDEC ID file now, even if the chip is not recognized (which happens to be the very reason I made that file...) Reading the IR cards may or may not have been fixed, sadly I do not have one at hand. Do you want to test it? If so, I will attach a couple of test builds, from two different commits. If not, I will have an IR cart myself in a week or two.

57fb0cd is the tip of dratini0:cartsave, 59c2f51 is the tip dratini0:cartsave_test, if you want to compile it yourself.

[GodMode9.zip](https://github.com/d0k3/GodMode9/files/3438770/GodMode9.zip)

**Comment by aspargas2** [comment]
The 57fb0cd build you posted and the 59c2f51 build you posted (which still identifies itself as `GodMode9 v1.8.0-93-g57fb0cd8` for some reason) appear to have the same behavior across the board:
* The system crashing when selecting `jedecid_and_sreg.bin` has been fixed for all carts I tested, but other strange issues have arisen with that file, described in the following points.
* For Pokemon Omega Ruby and Pokemon X, opening the file action menu on the `jedecid_and_sreg.bin` file or performing any operation on it (copying it somewhere, opening it in the hex editor, or calculating the SHA256) hangs the system for about one second, then appears to successfully perform the operation. However, calculating the SHA256 of the file directly, copying the file with the Y button to the SD then calculating the SHA256, and copying the file with the "copy to gm9/out/" option then calculating the SHA256 all yield different hashes, so something is obviously wrong with reading whatever data this file represents. Also, this file's hash directly in the virtual mount, hash after copying with the Y button, and hash after copying with the "copy to gm9/out/" option are each the same *between* Omega Ruby and X, which I don't think is supposed to happen. Only these two carts exhibited the hanging the system for one second behavior.
* For Pokemon Black and Black 2, the `.sav` file was still not present, and repeatedly calculating the SHA256 of `jedecid_and_sreg.bin` yielded a different hash every time. Different data was also present every time when repeatedly opening the hex editor for that file, and the byte 0xFF seemed to be commonly appearing.
* Everything else appears to be working the same as with 5b757f0.

**Comment by dratini0** [comment]
Thank you, that seems par for the course in most cases. The wrong commit hash in the build is because I built it before committing. So for each of your observations:

* Yay, I guess
* I forgot to add error handling code, thank you. If I did it right, it would be a read error or all just FFs.
* Okay, that is still pretty broken. I think we are at the point where I need to get my own IR cart, thank you for all your help!
* Good, thank you for testing those too!

**Comment by d0k3** [comment]
Alright, what's the state of this pull request @dratini0 ? Anything I can help with? Think it's ready to merge?

**Comment by dratini0** [comment]
Sorry for essentially disappearing for a while, I have been a bit busy. As things stand now, support for infrared carts is broken, and so is support for 8MB flash chips. I have had Art Academy for a while, (which is how I know it's broken), and I just got Active Health. (Curiously, it is broken in TWLSaveTool too)

I think I will have more to show in two weeks or so.

On 19 August 2019 21:10:18 BST, d0k3 <notifications@github.com> wrote:
>Alright, what's the state of this pull request @dratini0 ? Anything I
>can help with? Think it's ready to merge?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/d0k3/GodMode9/pull/517#issuecomment-522734326


Bálint Kovács

**Comment by d0k3** [comment]
@dratini0 - I'd suggest you just get back once you think it's ready to be reviewed and merged.

**Comment by d0k3** [comment]
@dratini0 - well, don't want to harass you, but any news? And, don't worry, it will take as long as you need. A small update on the current state of things would be nice, though (I really want this inside GM9, if that's not obvious).

**Comment by d0k3** [comment]
Oops, accidential close. Sorry!

**Comment by dratini0** [comment]
Hi!

I don't think I will get it finished to my complete satisfaction anytime soon. I have fixed 8MB games, like Art Academy, but I have been making little progress on IR titles. However, I did get a full logic analyzer trace from the actual hardware.

How about I just disable support for IR titles for now? That seems pretty mergable, and would take pressure off of me.

On 25 August 2019 19:21:37 BST, d0k3 <notifications@github.com> wrote:
>@dratini0 - I'd suggest you just get back once you think it's ready to
>be reviewed and merged.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/d0k3/GodMode9/pull/517#issuecomment-524652567


Bálint Kovács

**Comment by d0k3** [comment]
Yup, that's a pretty good solution. Fully acceptable. And, you may continue at any point in the future. Thanks a lot!

**Comment by dratini0** [comment]
Rebasing this onto staging was not my brightest moments, so I rebased it again against master. Should be ready for review and merge.

File: arm9/source/gamecart/gamecart.c [hunk 1]
```diff
@@ -221,3 +239,31 @@ u32 ReadCartPrivateHeader(void* buffer, u64 offset, u64 count, CartData* cdata)
...
+    u8 sReg;
+    if (offset >= JEDECID_AND_SREG_SIZE) return 1;
+    if (offset + count > JEDECID_AND_SREG_SIZE) count = JEDECID_AND_SREG_SIZE - offset;
+    int res = CardSPIReadJEDECIDAndStatusReg(cdata->save_type, &id, &sReg);
+    if (res) return res;
+    ownBuf[0] = (id >> 16) & 0xff;
+    ownBuf[1] = (id >> 8) & 0xff;
+    ownBuf[2] = id & 0xff;
+    ownBuf[JEDECID_AND_SREG_SIZE - 1] = sReg;
+    memcpy(buffer, ownBuf, count);
```
Comment (Wolfvak) [inline arm9/source/gamecart/gamecart.c]: Shouldn't this be `memcpy(buffer, ownBuf + offset, count);`? I know it'll be very rare to do offset reads on a (currently) 4 byte virtual file, but regardless this looks wrong.

Comment (dratini0) [reply arm9/source/gamecart/gamecart.c]: Didn't catch that, thank you!

**Comment by Wolfvak** [comment]
Okay, assuming all the magic hardware functions work correctly I'd say this is ready to merge, I've just reviewed it and other than a super minor issue, it looks fine.

**Comment by Wolfvak** [comment]
Last request from me - add yourself to the credits and list what exactly you implemented.

**Comment by d0k3** [comment]
Okay, just so everyone is on the same page: The code looks very good to me, and I'd have merged it by now, but there's one thing missing. I want someone but the dev (no offense meant, @dratini0, devs miss stuff), someone who preferable has a large amount of carts at their disposal (DS and 3DS) to test this before merging.

@TurdPooCharger maybe?

**Comment by dratini0** [comment]
Sorry, not yet. I think I can implement reading (since it is just the card's main memory), but I don't really have an idea on how to write them yet.

On 10 October 2019 19:23:01 BST, d0k3 <notifications@github.com> wrote:
>@dratini0 - a quick question - do you also handle card2 saves? Like
>Animal Crossing New Leaf? I'd just test, but I can't rn.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/d0k3/GodMode9/pull/517#issuecomment-540711026


Bálint Kovács

**Comment by TurdPooCharger** [comment]
@d0k3, sorry bud. None of my 3DS game collection are cartridges..

**Comment by d0k3** [comment]
@dratini0 don't worry too much. As long as it fails gracefully, it's all good.

**Comment by dratini0** [comment]
Yeah, it just classifies it as NO_CHIP, much like NTR "retail NAND" cards and other odd stuff.

On 11 October 2019 00:34:49 BST, d0k3 <notifications@github.com> wrote:
>@dratini0 don't worry too much. As long as it fails gracefully, it's
>all good.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/d0k3/GodMode9/pull/517#issuecomment-540840459


Bálint Kovács

**Comment by d0k3** [comment]
Okay, merged. Thank you, @dratini0! Looking forward to any other improvements you may add.

**Comment by d0k3** [comment]
Okay, we got a minor complaint. Pokemon X may be wrongly detected. There is a softlock when trying to open the "jedecid_and_sreg.bin" file. There is also a timeout here (arm9/source/gamecart/card_spi.c:126), but that takes 10 seconds, which is quite long. Would you recommend just shortening the delay, @dratini0? Maybe some better solution?

**Comment by dratini0** [comment]
Hmm, it might cause problems with writing Art Academy, but I will have to test that. If so, it might male sense to make the delay different for different operations.

What is X detected as?

On 14 October 2019 22:32:11 BST, d0k3 <notifications@github.com> wrote:
>Okay, we got a minor complaint. Pokemon X may be wrongly detected.
>There is a softlock when trying to open the "jedecid_and_sreg.bin"
>file. There is also a timeout here
>(arm9/source/gamecart/card_spi.c:126), but that takes 10 seconds, which
>is quite long. Would you recommend just shortening the delay,
>@dratini0? Maybe some better solution?
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/d0k3/GodMode9/pull/517#issuecomment-541934143


Bálint Kovács

**Comment by Wolfvak** [comment]
Sorry for the late reply, I'm the one who reported the Pokemon X issue. It's not that it's wrongly detected (I don't think it is, but how would I check the cart type anyway?), it's more that trying to open `jedecid_and_sreg.bin` times out the `CardSPIWaitWriteEnd` function.

**Comment by dratini0** [comment]
Okay, yes, that is fair enough - if we know for sure that a game has no SPI flash, there is no use in trying to find out what SPI flash it has. To get around this, I would need to add some logic to identify card2 and retail NAND games. (Also, sad times, card2 games are more complicated than I thought)

As for d0k3's idea of shortening the timeout it does seem to mess with 8MB writing, but it would still be beneficial for games (IIRC there are a handful) with no saving whatsoever.

I will come up with patches for these things Thursday-Friday-ish

**Pull request closed without merging**